v1/services/storage/store.go

```diff
@@ -650,27 +648,145 @@ func (s *Store) ReadSeriesCardinality(ctx context.Context, req *datatypes.ReadSe
+	if err != nil {
+		return nil, err
+	}
+
+	// Others use a slower way
+	c2, err := s.seriesCardinalityWithPredicateAndTime(ctx, s.TSDBStore.Shards(shardsPartiallyInTimeRange), expr, sfile, start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	ss := tsdb.NewSeriesIDSet()
+	ss.Merge(c1, c2)
+
+	return cursors.NewInt64SliceIterator([]int64{int64(ss.Cardinality())}), nil
+}
+
+func (s *Store) seriesCardinalityWithPredicate(ctx context.Context, shards []*tsdb.Shard, expr influxql.Expr, sfile *tsdb.SeriesFile) (*tsdb.SeriesIDSet, error) {
+	if expr == nil {
+		return s.TSDBStore.SeriesCardinalityFromShards(ctx, shards)
 	}
-	if req.Range.End == 0 {
-		req.Range.End = models.MaxNanoTime
+
+	ss := tsdb.NewSeriesIDSet()
+	if len(shards) == 0 {
+		return ss, nil
 	}
 
-	database, _, _, _, err := s.validateArgs(source.OrgID, source.BucketID, req.Range.Start, req.Range.End)
+	cur, err := newIndexSeriesCursorInfluxQLPred(ctx, expr, shards)
 	if err != nil {
 		return nil, err
 	}
 
-	card, err := s.TSDBStore.SeriesCardinality(ctx, database)
+	buf := make([]byte, 1024)
+	for {
+		r := cur.Next()
+		if r == nil {
+			break
+		}
+		skey := sfile.SeriesID(r.Name, r.SeriesTags, buf)
+		ss.Add(skey)
+	}
+
+	return ss, nil
+}
+
+func (s *Store) seriesCardinalityWithPredicateAndTime(ctx context.Context, shards []*tsdb.Shard, expr influxql.Expr, sfile *tsdb.SeriesFile, start, end int64) (*tsdb.SeriesIDSet, error) {
+	ss := tsdb.NewSeriesIDSet()
+	if len(shards) == 0 {
+		return ss, nil
+	}
+
+	cur, err := newIndexSeriesCursorInfluxQLPred(ctx, expr, shards)
 	if err != nil {
 		return nil, err
 	}
 
-	return cursors.NewInt64SliceIterator([]int64{card}), nil
+	buf := make([]byte, 1024)
+	rs := reads.NewFilteredResultSet(ctx, start, end, cur)
+	for rs.Next() {
+		func() {
+			c := rs.Cursor()
+			if c == nil {
+				// no data for series key + field combination
+				return
+			}
+			defer c.Close()
+
+			if cursorHasData(c) {
+				r := cur.row
+				skey := sfile.SeriesID(r.Name, r.SeriesTags, buf)
+				ss.Add(skey)
+			}
+		}()
+	}
+
+	return ss, nil
 }
 
 func (s *Store) SupportReadSeriesCardinality(ctx context.Context) bool {
-	// Disabled until predicate filtering for cardinality is fully implemented.
-	return false
+	return true
+}
+
+// Returns two slices of shard IDs - the first is shards that are entirely in
+// the provided time range; the second is shards that are not entirely within
+// the provided time range.
+func groupShardsByTime(sgs []meta.ShardGroupInfo, start, end int64) ([]uint64, []uint64) {
+	entirelyInRange := []uint64{}
+	partiallyInRange := []uint64{}
+
+	for _, sg := range sgs {
+		shards := make([]uint64, 0, len(sg.Shards))
+		for _, si := range sg.Shards {
+			shards = append(shards, si.ID)
+		}
+
+		// shard is entirely within the specified time range
+		if sg.StartTime.After(time.Unix(0, start)) && sg.EndTime.Before(time.Unix(0, end)) {
```

Comment:
Are `After` and `Before` inclusive? Or exclusive?

Reply:
Exclusive! Nice catch. I updated the code to check for shards being within the range inclusive and added some test cases for this method.